criu/files-reg.c

```diff
@@ -1581,6 +1587,110 @@ static int get_build_id(const int fd, const struct stat *fd_status,
+		*iter = 0;
+}
+
+static inline void calculate_checksum_iterator_next(u64 *iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FULL)
+		*iter += 1;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
+		*iter += 1;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD)
+		*iter += opts.file_validation_chksm_parameter;
+}
+
+static inline bool calculate_checksum_iterator_stop_condition(u64 iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FULL) {
+		return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST) {
+		if (iter >= opts.file_validation_chksm_parameter)
+			return true;
+		else
+			return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Does the actual work of calculating the CRC32C checksum of "some"
+ * parts of the file. These "some" parts depend on the configuration chosen.
+ * By default, the first 1024 bytes of the file or the entire file, which ever
+ * is smaller, is considered.
+ * Returns true if the checksum could be obtained, else it returns false.
+ */
+static bool calculate_checksum(const int fd, const struct stat *fd_status,
+				u32 *checksum)
+{
+	int tmp;
+	u32 byte, mask;
+	u64 up_bound, low_bound = 0, i = 0;
+	unsigned char *file_header;
+
+	/* At most, the first 10MB of the file is mapped. */
+	up_bound = min_t(size_t, fd_status->st_size, CHKSM_CHUNK_SIZE);
+	file_header = (unsigned char *) mmap(0, up_bound,
+						PROT_READ, MAP_PRIVATE | MAP_FILE, fd, 0);
+	if (file_header == MAP_FAILED) {
+		pr_warn("Couldn't mmap file with fd %d", fd);
+		return false;
+	}
+
+	*checksum = 0xFFFFFFFF;
+	calculate_checksum_iterator_init(&i);
+
+	while (!calculate_checksum_iterator_stop_condition(i) &&
+			i >= low_bound && i < up_bound) {
+
+		byte = file_header[i-low_bound];
+		*checksum = *checksum ^ byte;
+		for (tmp = 0; tmp < 8; tmp++) {
+			mask = -(*checksum & 1);
+
+			/*
+			 * Little endian notation is used.
+			 * The Castagnoli polynomial (0x82F63B78) is used instead of
+			 * 0xEDB88320 and is the difference between CRC32C and CRC32.
+			 */
+			*checksum = (*checksum >> 1) ^ (0x82F63B78 & mask);
+		}
+
+		calculate_checksum_iterator_next(&i);
+
+		/*
+		 * If the new iterator position is outside the mapped region (Bytes in the range
+		 * low_bound and up_bound) of the file, the current region needs to be unmapped
+		 * and the next portion (At most the next 10MB) of the file needs to be mapped.
+		 */
+		if (i >= up_bound && up_bound != fd_status->st_size && i < fd_status->st_size) {
+			munmap(file_header, up_bound - low_bound);
+			low_bound = round_down(i, sysconf(_SC_PAGE_SIZE));
+			up_bound = min_t(size_t, fd_status->st_size, low_bound + CHKSM_CHUNK_SIZE);
+			file_header = (unsigned char *) mmap(0, up_bound - low_bound,
```

Comment:
This is not a file header. I think you need to rename it to "buf" or something like this.